fluster/decoders/gstreamer.py

```diff
@@ -71,10 +72,56 @@ def __init__(self) -> None:
         if not gst_element_exists(self.sink):
             self.sink = 'filesink'
 
-    def gen_pipeline(self, input_filepath: str, output_filepath: str, output_format: OutputFormat) -> str:
+    def gen_pipeline(self, input_filepath: str, output_filepath: Optional[str], output_format: OutputFormat) -> str:
         '''Generate the GStreamer pipeline used to decode the test vector'''
         # pylint: disable=unused-argument
-        return PIPELINE_TPL.format(self.cmd, input_filepath, self.decoder_bin, self.caps, self.sink, output_filepath)
+        output = "location={}".format(output_filepath) if output_filepath else ""
+        return PIPELINE_TPL.format(self.cmd, input_filepath, self.decoder_bin, self.caps, self.sink, output)
+
+    def parse_md5sum(self, data: Tuple[str, str], verbose: bool) -> str:
+        '''Parse the MD5 sum out of commandline output'''
+        md5sum = "error"
+        for line in filter(None, data):
+            if verbose:
+                print(line, end='')
+            pattern = "conformance/checksum, checksum-type=(string)MD5, checksum=(string)"
+            sum_start = line.find(pattern)
+            if sum_start > 0:
+                sum_start += len(pattern)
+                sum_end = line[sum_start:].find(";")
+                if sum_end > 0:
+                    sum_end += sum_start
+                    md5sum = line[sum_start:sum_end]
+        return md5sum
+
+    def decode2(
```

Comment:
I think this is not a good function name, I think it is confusing.
From my point of view the functions do the following:
decode: decode the byte stream to a file and get the checksum from the file
decode2: decode the byte stream and read the checksum from STDOUT (write to the file optionally)

So, I have two alternatives in mind to improve this:
1. Rename the function: `decode2 -> read_checksum_from_stdout`
2. Just extend the `decode` function similar to this, as the function is quite short currently and adding a single branch doesn't make it harder to read.
```
    def decode(
        self,
        input_filepath: str,
        output_filepath: str,
        output_format: OutputFormat,
        timeout: int,
        verbose: bool,
        keep_files: bool = False,
    ) -> str:
        '''Decode the test vector and do the checksum'''
        # Read the checksum from STDOUT
        if self.sink == 'videocodectestsink' and not keep_files:
            pipeline = self.gen_pipeline(input_filepath, None, output_format)
            command = shlex.split(pipeline)
            command.append("-m")
            serr = subprocess.DEVNULL if not verbose else None
            if verbose:
                print(f'\nRunning command "{" ".join(command)}"')

            try:

                with subprocess.Popen(command, stdout=subprocess.PIPE,
                                      stderr=serr, universal_newlines=True) as pipe:
                    data = pipe.communicate(timeout=timeout)
                    return self.parse_md5sum(data, verbose)
            except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as ex:
                # Developer experience improvement (facilitates copy/paste)
                ex.cmd = " ".join(ex.cmd)
                raise ex
        # Read the checksum from the produced file
        pipeline = self.gen_pipeline(input_filepath, output_filepath, output_format)
        run_command(shlex.split(pipeline), timeout=timeout, verbose=verbose)
        return file_checksum(output_filepath)
```

Notice also the: `if self.sink == 'videocodectestsink' and not keep_files:`, looking at it I feel like we only need to read from STDOUT, when we don't keep the file, so if we have the `videocodectestsink` but we still write a file we can go the old route, right?

Reply:
I also don't like `decode2` name.

Reply:
We can go the old route yes, but that will be slower (1 full extra raw data read).

Reply:
I have squashed the decode2 into the decode function as suggested (avoiding to have to pick a better name). I've added a comment on why we spacial case videocodectestsink, and kept the optimization for the keep_file mode, since over NFS I do see a appreciable difference in avoiding the extra read back (perhaps my cache is too small, but that's just another subject).